### PR TITLE
Optimize series rolling count

### PR DESCRIPTION
**Pull request by Rubtsowa**

### CODE REVIEW

**Comment by AlexanderKalistratov** [comment]
Can't this be based on recently merged rolling.sum?

**Comment by Rubtsowa** [comment]
name | nthreads | type | size | median | min | max | compile | boxing
-- | -- | -- | -- | -- | -- | -- | -- | --
Series.rolling.count | 1 | Python | 10000000 | 3.279528 | 3.261298 | 3.300166 |   |  
Series.rolling.count | 1 | SDC | 10000000 | 14.15811 | 14.14303 | 14.19892 | 0.4142 | 0.001624
  |   |   |   |   |   |   |   |  
name | nthreads | type | size | median | min | max | compile | boxing
Series.rolling.count | 1 | Python | 10000000 | 2.486854 | 2.360286 | 2.515337 |   |  
Series.rolling.count | 1 | SDC | 10000000 | 0.136682 | 0.121665 | 0.140139 | 0.167185 | 0.001064

**Comment by Rubtsowa** [comment]
median       min       max   compile  boxing
name                 nthreads type   size
Series.rolling.count 1        Python 10000000  1.303667  1.281662  1.504678       NaN     NaN
                              SDC    10000000  0.099000  0.085000  0.131000  0.216372     0.0
Series.rolling.count 2        Python 10000000  1.526775  1.482826  1.589015       NaN     NaN
                              SDC    10000000  0.090000  0.085000  0.100000  0.291824     0.0
Series.rolling.count 3        Python 10000000  1.294494  1.288564  1.519767       NaN     NaN
                              SDC    10000000  0.085000  0.084000  0.100000  0.216333     0.0
Series.rolling.count 4        Python 10000000  1.296684  1.272419  1.366641       NaN     NaN
                              SDC    10000000  0.085000  0.084000  0.085000  0.216261     0.0

**Comment by Rubtsowa** [comment]
name | nthreads | type | size | median | min | max | compile | boxing
-- | -- | -- | -- | -- | -- | -- | -- | --
Series.rolling.count | 1 | Python | 10000000 | 1.303667 | 1.281662 | 1.504678 | NaN | NaN
  |   | SDC | 10000000 | 0.099 | 0.085 | 0.131 | 0.216372 | 0
  | 2 | Python | 10000000 | 1.526775 | 1.482826 | 1.589015 | NaN | NaN
  |   | SDC | 10000000 | 0.09 | 0.085 | 0.1 | 0.291824 | 0
  | 4 | Python | 10000000 | 1.296684 | 1.272419 | 1.366641 | NaN | NaN
  |   | SDC | 10000000 | 0.085 | 0.084 | 0.085 | 0.216261 | 0

**Comment by AlexanderKalistratov** [comment]
Please rebase on master & replace `sdc_rolling_overload` by `sdc_method_overload` for count

**Comment by AlexanderKalistratov** [comment]
And then remeasure results

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py [hunk 1]
```diff
@@ -305,6 +305,32 @@ def apply_minp(arr, ddof, minp):
     gen_hpat_pandas_series_rolling_ddof_impl(arr_var))
 
 
+@sdc_register_jitable
+def pop_count(value, nfinite, result):
+    """Calculate the window sum without old value."""
+    if not numpy.isnan(value):
+        nfinite -= 1
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_series_rolling_functions.py]: I don't think we need to do something with `nfinite`. Could you please remove it?

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py [hunk 2]
```diff
@@ -305,6 +305,32 @@ def apply_minp(arr, ddof, minp):
...
+        result -= 1
+
+    return nfinite, result
+
+
+@sdc_register_jitable
+def put_count(value, nfinite, result):
+    """Calculate the window sum with new value."""
+    if not numpy.isnan(value):
+        nfinite += 1
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_series_rolling_functions.py]: and here too

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py [hunk 3]
```diff
@@ -355,6 +381,13 @@ def impl(self):
         length = len(input_arr)
         output_arr = numpy.empty(length, dtype=float64)
 
+        if minp == 0 and win == 0:
+            for i in prange(length):
+                output_arr[i] = init_result
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_series_rolling_functions.py]: I'm not sure... shouldn't it be something like:
```python
output_arr[i] = get_result(nfinite, minp, init_result)
```

@densmirn

Comment (AlexanderKalistratov) [reply sdc/datatypes/hpat_pandas_series_rolling_functions.py]: BTW is it correct behavior at all - create rolling with minp == 0 and win == 0?

**Comment by Rubtsowa** [comment]
name | nthreads | type | size | median | min | max | compile | boxing
-- | -- | -- | -- | -- | -- | -- | -- | --
Series.rolling.count | 1 | Python | 10000000 | 1.431035 | 1.385032 | 1.49204 | NaN | NaN
  |   | SDC | 10000000 | 0.101 | 0.099 | 0.117 | 0.310997 | 0.000003
  | 4 | Python | 10000000 | 1.710061 | 1.389035 | 1.802042 | NaN | NaN
  |   | SDC | 10000000 | 0.035 | 0.032 | 0.035 | 0.339018 | 0.000002

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py [hunk 4]
```diff
@@ -305,6 +305,30 @@ def apply_minp(arr, ddof, minp):
     gen_hpat_pandas_series_rolling_ddof_impl(arr_var))
 
 
+@sdc_register_jitable
+def pop_count(value, nfinite, result):
+    """Calculate the window sum without old value."""
+    if not numpy.isnan(value):
+        result -= 1
+
+    return nfinite, result
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_series_rolling_functions.py]: I would implement it something like this:
```python
if numpy.isnan(value):
    return nfinite, result

return nfinite, result - 1
```

The same is applicable for `put_count`.

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py [hunk 5]
```diff
@@ -305,6 +305,30 @@ def apply_minp(arr, ddof, minp):
...
+def put_count(value, nfinite, result):
+    """Calculate the window sum with new value."""
+    if not numpy.isnan(value):
+        result += 1
+
+    return nfinite, result
+
+
+@sdc_register_jitable
+def result_count(nfinite, minp, result):
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_series_rolling_functions.py]: Maybe `count_result` as for method `mean`?

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py (hunk 3 above)
Comment (densmirn) [reply sdc/datatypes/hpat_pandas_series_rolling_functions.py]: @AlexanderKalistratov it's correct from Pandas side, but strange in general.

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py [hunk 6]
```diff
@@ -388,6 +419,8 @@ def impl(self):
     return impl
 
 
+sdc_pandas_series_rolling_count_impl = register_jitable(
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_series_rolling_functions.py]: `register_jitable` is excess here.

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py (hunk 3 above)
Comment (densmirn) [reply sdc/datatypes/hpat_pandas_series_rolling_functions.py]: @Rubtsowa the extension looks strange. The old code didn't have such checks.

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py (hunk 5 above)
Comment (Rubtsowa) [reply sdc/datatypes/hpat_pandas_series_rolling_functions.py]: No, here is different.

Comment (densmirn) [reply sdc/datatypes/hpat_pandas_series_rolling_functions.py]: Let it be.

Comment (densmirn) [reply sdc/datatypes/hpat_pandas_series_rolling_functions.py]: Let's generalize the function and name it just `result` or `get_result` without connections to the method `count`. Don't forget to change docstring. BTW I see mention about min periods in the docstring, but don't see usage of `minp` in the function.

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py (hunk 3 above)
Comment (AlexanderKalistratov) [reply sdc/datatypes/hpat_pandas_series_rolling_functions.py]: Do we have tests for this, btw?

Comment (densmirn) [reply sdc/datatypes/hpat_pandas_series_rolling_functions.py]: @AlexanderKalistratov yes:
```python
for window in range(0, len(obj) + 3, 2):
    for min_periods in range(0, window + 1, 2):
        # ...
```

Comment (AlexanderKalistratov) [reply sdc/datatypes/hpat_pandas_series_rolling_functions.py]: @Rubtsowa so, count fails this tests without the fix?

**Comment by Rubtsowa** [comment]
name                     nthreads   type      size      median       min          max      compile    boxing
                          
Series.rolling.count 1        Python 10000000  1.310429  1.272626  1.424345       NaN       NaN
                                           SDC    10000000  0.085000  0.078000  0.100000  0.231873  0.000145
Series.rolling.count 4        Python 10000000  1.37744  1.335639  1.479352       NaN       NaN
                                          SDC    10000000  0.04900  0.040000  0.051000  0.273639  0.000511

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py [hunk 7]
```diff
@@ -496,32 +522,13 @@ def calc_corr(main, other, minp):
...
-        output_arr = numpy.empty(length, dtype=float64)
-
-        boundary = min(win, length)
-        for i in prange(boundary):
-            arr_range = input_arr[:i + 1]
-            output_arr[i] = arr_nonnan_count(arr_range)
-
-        for i in prange(boundary, length):
-            arr_range = input_arr[i + 1 - win:i + 1]
-            output_arr[i] = arr_nonnan_count(arr_range)
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_series_rolling_functions.py]: `arr_nonnan_count` is not used at all. So let's also remove this function.

**Comment by densmirn** [comment]
> name	nthreads	type	size	median	min	max	compile	boxing
> Series.rolling.count	1	Python	10000000	1.310429	1.272626	1.424345	NaN	NaN
> SDC	10000000	0.085000	0.078000	0.100000	0.231873	0.000145
> Series.rolling.count	4	Python	10000000	1.37744	1.335639	1.479352	NaN	NaN
> SDC	10000000	0.04900	0.040000	0.051000	0.273639	0.000511

Result numbers are too small. Let's increase number of input data.

File: sdc/datatypes/hpat_pandas_series_rolling_functions.py [hunk 8]
```diff
@@ -362,10 +386,10 @@ def impl(self):
             result = init_result
 
             prelude_start = max(0, chunk.start - win + 1)
-            prelude_stop = min(chunk.start, prelude_start + win)
+            prelude_stop = chunk.start
 
             interlude_start = prelude_stop
-            interlude_stop = min(prelude_start + win, chunk.stop)
+            interlude_stop = min(interlude_start + win, chunk.stop)
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_series_rolling_functions.py]:
```suggestion
            interlude_stop = min(prelude_start + win, chunk.stop)
```

Comment (densmirn) [reply sdc/datatypes/hpat_pandas_series_rolling_functions.py]: Don't apply suggested change. Perhaps that's incorrect.

**Comment by Rubtsowa** [comment]
name | nthreads | type | size | median | min | max | compile | boxing
-- | -- | -- | -- | -- | -- | -- | -- | --
Series.rolling.count | 1 | Python | 10000000 | 0.810121 | 0.807817 | 0.81306 |   |  
Series.rolling.count | 1 | SDC | 10000000 | 0.068798 | 0.068041 | 0.068811 | 0.234757 | 0.000512
Series.rolling.count | 2 | Python | 10000000 | 0.753858 | 0.750156 | 0.767349 |   |  
Series.rolling.count | 2 | SDC | 10000000 | 0.034607 | 0.03448 | 0.035149 | 0.233084 | 0.000527
Series.rolling.count | 4 | Python | 10000000 | 0.720124 | 0.717565 | 0.72268 |   |  
Series.rolling.count | 4 | SDC | 10000000 | 0.017625 | 0.017254 | 0.018111 | 0.224228 | 0.000452
Series.rolling.count | 8 | Python | 10000000 | 0.716457 | 0.714138 | 0.716962 |   |  
Series.rolling.count | 8 | SDC | 10000000 | 0.009069 | 0.008596 | 0.009097 | 0.226648 | 0.000489
Series.rolling.count | 16 | Python | 10000000 | 0.714823 | 0.712627 | 0.717463 |   |  
Series.rolling.count | 16 | SDC | 10000000 | 0.009171 | 0.008184 | 0.009809 | 0.22575 | 0.000432
Series.rolling.count | 28 | Python | 10000000 | 0.72093 | 0.716412 | 0.722183 |   |  
Series.rolling.count | 28 | SDC | 10000000 | 0.006558 | 0.005482 | 0.007863 | 0.227914 | 0.00049
Series.rolling.count | 56 | Python | 10000000 | 0.720681 | 0.705925 | 0.730319 |   |  
Series.rolling.count | 56 | SDC | 10000000 | 0.005491 | 0.004581 | 0.006185 | 0.231097 | 0.000432

**Comment by AlexanderKalistratov** [comment]
conflict

**Pull request merged**